Pallet Staking Lottery

Garandor opened the pull request:
Implements an APY-maximizing no-loss lottery for Manta/Calamari

- `liquidate_lottery` and `rebalance_stake` extrinsics are left unimplemented for later
- [x] reverify `DrawingFreezeout` is long enough for both Polkadot and Kusama

Note that the test `winner_distribution_should_be_equal_with_equal_deposits` is nondeterministic and can sometimes fail.
I have set the boundary low enough for random failures to be rare, but nonetheless, this test failing is not necessarily indication of an error if it occurs  isolated.

Signed-off-by: Adam Reif <Garandor@manta.network>

---

Before we can approve this PR for merge, please make sure that **all** the following items have been checked off:
- [ ] Connected to an issue with discussion and accepted design using zenhub "Connect issue" button below
- [x] Added **one** label out of the `L-` group to this PR
- [x] Added **one or more** labels from the `A-` and `C-` groups to this PR
- [x] Explicitly labelled `A-calamari`, `A-dolphin` and/or `A-manta` if your changes are meant for/impact either of these (CI depends on it)
- [ ] Re-reviewed `Files changed` in the Github PR explorer.


Situational Notes:
- If adding functionality, write unit tests!
- If importing a new pallet, choose a proper module index for it, and allow it in `BaseFilter`. Ensure **every** extrinsic works from front-end. If there's corresponding tool, ensure both work for each other.
- If needed, update our Javascript/Typescript APIs. These APIs are officially used by exchanges or community developers.
- If modifying existing runtime storage items, make sure to implement storage migrations for the runtime and test them with `try-runtime`. This includes migrations inherited from upstream changes, and you can search the diffs for modifications of `#[pallet::storage]` items to check for any.

Commit:
WIP: lottery pallet

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
wip: First impl of

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
wip: refactor unstaking/staking of collators instead of per-user-request

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
wip: syntax fixes

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
wip: weight

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
wip: more syntax fixes

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
further typing fixes

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
more type and origin fixes

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
next batch of fixes

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
comment drawing scheduling

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix randomness conversion to Balance

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
remove the alice collator crutch, now selects first active collator

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
more fixes

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
wip: first batch move semantics

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
pallet compiles :tada:

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
add lottery pallet to calamari

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Merge branch 'manta' into garandor/pt

Commit:
revert changes to manta RT

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix lottery on calamari

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
remove frame-executive

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix wasm/node compilation

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
clean dupe

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
unfilter lottery

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Council can run Scheduler

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
wip: schedule

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
wip: create expected calltype

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
trait bounds for `Call`

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fmt

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
wip: loose coupled scheduler

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Use Currency from staking, get rid of related trait bounds

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
wip: try not defining a custom Call in pallet and use frame_system::Call directly

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
wip

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fixed Call conversion

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
prod_or_fast! for lottery and set next drawing on start

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
DrawingInterval DrawingFreezeout UnstakeLockTime as constants plus other changes

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
manual winnings claim + sanity and lower bound checks

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
track total winnings correctly. add extrinsic to manually process withdrawals and liquidate the lottery

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
wip

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Merge v4.0.4

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
doc: Fix broken links

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
doc: lottery rustdoc

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
doc: remove origin param

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
doc: more

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
doc: cleanup TODOs

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
doc: mention pdot.js for state querying

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
simplify ensure_root_or_manager to just the ManageOrigin, it's set in the runtime

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
disable root-only

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix lottery surplus calculation

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
remove NextDrawingAt storage, logic fixes

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
logging

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Revert "logging"

This reverts commit 9577bbaaaf8284b9c3a805faf096dc158f2cf20d.

Commit:
Revert "disable root-only"

This reverts commit 8e01020165e270a9eeec2f3711e656da50f8e1d9.

Commit:
introduce lottery_schedule_id helper fn

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
extra doc, split pub/priv fns

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fork pallet randomness from moonbeam 41c0131e72b628418ff1dc4ede0ca8810bf250fd

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
add inherent from the same commit, change L57 `asnyc fn` to `fn`

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
remove all evm and localVRF stuff. pallet_randomness lib.rs compiles

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fmt

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Add pallet_randomness to calamari RT, remove Currency trait

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix no_std build

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fmt

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
unfilter set_babe_randomness_results inherent

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
cleanup

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fmt

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
REVERT THIS disable Manager check

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
use two-epoch-ago for randomness

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
wip: Deposit distribution to multiple collators: re-activate collators we fell out of the active set with

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
wip: prefer staking to bottom, unstake from inactive collators

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
finish up deposit distribution logic

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
finish withdrawal selection logic

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
enable collator selection in `request_withdraw`

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
update SumOfDeposits when excuting unstaking and add collator selection to deposits

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
skip to random assignment if there are no underallocated collators. add debug logs

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
logic fix in start_lottery

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
reenable check if freezeout period was set correctly

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
more debug traces

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
add gas fee checks and estimation for deposits/withdrawals

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
two logic fixes

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
mock runtime and first unit test

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Revert "REVERT THIS disable Manager check"

This reverts commit a01386a121dbe58176a8edaaedf62c57f034306f.

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
move `stake_collator` and `unstake_collator` to `staking` module

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
first tests and logic fixes

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Merge branch 'demo' into garandor/pt

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
start/stop, basic deposit/withdraw working

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Revert "Revert "REVERT THIS disable Manager check""

This reverts commit 0774a33b037f249cd1eb009b8f27667b0cf197e2.

Commit:
fix backwards withdrawal check logic, also fmt

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
add function_name crate for tracing

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix: Don't pay withdrawal requests that are still timelocked

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix: deposit calculation underflow

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
change: deposits underallocated now counts at-or-below-mean stake

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
refactor: winner selection is its own fn

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Fix: Underallocation calculation for deposits

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix: remove keys from storageMaps when value is deleted

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
reenable Manage origin check

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Merge tag 'v4.0.6' into garandor/pt

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix: compilation with try_mutate_exists

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fmt

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix: zero-division in drawing call

Signed-off-by: Adam Reif <Garandor@manta.network>

ferrell-code committed:
use older sr-tool version

Signed-off-by: Charles Ferrell <charlie@manta.network>

Commit:
Merge tag 'v4.0.7' into garandor/pt

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
node script to start lottery on baikal

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Merge tag 'v4.0.7' into jumboshrimps

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Merge tag 'v4.0.8' into jumboshrimps

Manta/Calamari v4.0.8 Release

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Expose lottery state queries via RPC. Rename some stuff

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Merge tag 'v4.0.8-1' into jumboshrimps

Manta v4.0.8-1 hot fix

Commit:
reenable logic to not count outstanding_withdrawal_requests to prize_pool

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix pallet_randomness tests and benchmarks

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix pallet-lottery tests

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
use blocknumber instead of total pot for fallback collator selection

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fmt

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
clippy cleanup

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
add first lottery benchmark, fix pallet-randomness benchmark

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
lottery: `deposit` benchmark, enable benchmarking

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
minor: reorder

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
minor: cleanup unused

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix non-benchmark

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
bench: request_withdraw

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
add claim_my_winnings benchmark and fix tests. allow manual `draw_lottery` even when lottery is stopped

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
prevent crash when no collators found
add test depositing_and_withdrawing_in_freezeout_should_not_work

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
seed fallback randomness from block number and extrinsic index

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
cleanup

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Merge branch 'manta' into jumboshrimps

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix removed include

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Enable lottery pallet on manta runtime

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fail rebalance and liquidate extrinsics

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
clippy

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
add total_users storage item and weights files

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
clippy

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
remove modulo bias from lottery drawing

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Merge branch 'manta' into garandor/pt

Commit:
test winning amount = 1 to see number of wins easier

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
cleanup mean stake no longer used

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
remaning todos for v1

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Merge branch 'manta' into garandor/pt

Commit:
unwrap -> expect

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
more clippy

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix try-runtime

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
also for calamari

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
more clippy

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
more clippy

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
more clippy

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
more clippy

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Starded/Stopped event without param, even more clippy, lottery stop test

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
more clippy

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Merge branch 'manta' into garandor/pt

Commit:
minor: remove unused macro

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix parachain-staking benchmark

Signed-off-by: Adam Reif <Garandor@manta.network>

ghzlatarev commented on pallets/randomness/src/lib.rs:
```diff
@@ -0,0 +1,201 @@
+// Copyright 2019-2022 PureStake Inc.
+// This file is part of Moonbeam.
```
moonbeam or manta here ?

Garandor replied:
substituted with manta headers. resolving

ghzlatarev commented on pallets/randomness/src/lib.rs:
```diff
@@ -0,0 +1,201 @@
+                RandomnessResults::<T>::get(RequestType::BabeEpoch(relay_epoch_index))
+                    .unwrap()
+                    .randomness
+                    .unwrap(); // TODO: Handle unavailable randomness
```
A bunch of TODOs in the code. Not sure which of them are supposed to be done now and which are issues for the future.

Garandor replied:
thanks, the randomness ones were actually important. Although these unwraps should be impossible to get triggered.

ghzlatarev commented on pallets/pallet-lottery/src/lib.rs:
```diff
@@ -0,0 +1,1074 @@
+    #[pallet::storage]
+    #[pallet::getter(fn withdrawal_request_queue)]
+    pub(super) type WithdrawalRequestQueue<T: Config> =
+        StorageValue<_, Vec<Request<T::AccountId, T::BlockNumber, BalanceOf<T>>>, ValueQuery>;
```
Maybe this should be bounded vec ? Someone could do a bunch of min-value requests to spam

ghzlatarev replied:
I guess it can also grow infinitely since there's no time limit to withdrawing

Garandor replied:
These requests are time-bounded because they will be paid out after timelock expiry, no way for users to prevent that other than by stopping the lottery or if drawings fail to execute.
The anti-spam bound on this is `min_withdraw` again so it needs to be reasonably set

ghzlatarev replied:
oof yeah of course, a user requests a withdraw which fills up this queue, and it gets freed up after however many blocks, so it can't grow infinitely.

ghzlatarev commented on pallets/pallet-lottery/src/staking.rs:
```diff
@@ -0,0 +1,495 @@
+        Ok(())
+    }
+
+    #[named]
```
why was this attribute needed btw ?

Garandor replied:
for `log::trace!(function_name!());`

ghzlatarev commented on pallets/pallet-lottery/src/staking.rs:
```diff
@@ -0,0 +1,495 @@
+                    // being selected is not enough, it must also be actively participating in block production
+                    // but we need to ensure it had a chance to produce blocks, so we only check this if we're at least 25% into a round
+                    if now < info.first + (info.length / 4u32).into() {
+                        return true;
```
Does the logic rely on each collator producing at least one block ? The more the better ? Or that you will keep producing blocks ?

Garandor replied:
only collators that have produced at least 1 block are considered active.
unless we're less than 25% into a round, then *all* selected collators are considered active.

ghzlatarev replied:
aha ok i get it now, but isn't 1/4 of a round too long . Can't we use the number of the candidates amount of blocks into the round ?

Garandor replied:
people can randomly drop blocks sometimes, no need to unstake if they missed their first block or two for whatever reason

Garandor replied:
discussion is obsolete, i changed it to check if this collator has earned any points in the last round instead

ghzlatarev commented on pallets/pallet-lottery/src/staking.rs:
```diff
@@ -0,0 +1,495 @@
+            pallet_parachain_staking::Pallet::<T>::compute_top_candidates(); // XXX: This can select collators that are joined but not yet producing blocks
+        if top_collator_accounts.is_empty() {
+            // Use `SelectedCandidates` as fallback (should basically never happen)
+            log::warn!("Lottery falling back to selected candidates");
```
maybe put a why this is a warning in the message

Garandor replied:
done

ghzlatarev commented on pallets/pallet-lottery/src/staking.rs:
```diff
@@ -0,0 +1,495 @@
+        // TODO: Small optimization: Also consider points / pointsAwarded to not stake to collators missing blocks
+
+        let mut top_collator_accounts =
+            pallet_parachain_staking::Pallet::<T>::compute_top_candidates(); // XXX: This can select collators that are joined but not yet producing blocks
```
why is the comment important

ghzlatarev replied:
also maybe it would be better to just divide evenly in v1, this adds a lot of complexity

Garandor replied:
> why is the comment important

because the goal here is APY maximization. It's a minor point though so i left it out of scope.

> also maybe it would be better to just divide evenly in v1, this adds a lot of complexity

i agree, but that complexity is contained and largely provided by parachain_staking.
Open to removing it, but that would also remove one of the main selling points and thus becomes *a product decision*

Commit:
Merge branch 'manta' into jumboshrimps

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
disable freezeout sanity check

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fixup

Signed-off-by: Adam Reif <Garandor@manta.network>

Dengjianping commented on pallets/pallet-lottery/src/lib.rs:
```diff
@@ -0,0 +1,1077 @@
+
+    #[pallet::pallet]
+    #[pallet::storage_version(STORAGE_VERSION)]
+    // #[pallet::generate_store(pub(super) trait Store)]
```
Removable?

Garandor replied:
done thx

Dengjianping commented on pallets/pallet-lottery/src/lib.rs:
```diff
@@ -0,0 +1,1077 @@
+    pub enum Event<T: Config> {
+        LotteryStarted,
+        LotteryStopped,
+        LotteryWinner(T::AccountId, BalanceOf<T>),
```
Why not to use structured events?

ghzlatarev replied:
what do you mean exactly ?

Dengjianping replied:
Like
```rust
LotteryWinner { winner: T::AccountId, amount: BalanceOf<T> }
```

Garandor replied:
cool thx @Dengjianping, i wasn't aware that syntax was possible. Changed

Commit:
Merge branch 'manta' into garandor/pt

Dengjianping commented:
I doubt whether we need customized rpc methods, there's `state_call`, once runtime apis are defined and implemented(already done), we can call runtime apis by `state_call`, and no need to maintain those rpc anymore.

But if we concern about the performance of rpc method, like mantapay's rpc methods, we can still stick to native rpc implementation.

Dengjianping commented on pallets/pallet-lottery/Cargo.toml:
```diff
@@ -0,0 +1,78 @@
+authors = ["Adam Reif"]
+edition = "2021"
+name = "pallet-lottery"
+version = "0.1.0"
```
version number

Dengjianping replied:
And more metadata about this pallet

Garandor replied:
the code was originally forked from their 0.9.27 codebase, updated

Dengjianping commented on pallets/pallet-lottery/Cargo.toml:
```diff
@@ -0,0 +1,78 @@
+version = "0.1.0"
+
+[dependencies]
+codec = { version = '3.1.2', default-features = false, features = ['derive'], package = 'parity-scale-codec' }
```
seems pretty old version

Garandor replied:
true, updated it

Dengjianping commented on pallets/pallet-lottery/src/lib.rs:
```diff
@@ -0,0 +1,1077 @@
+
+    #[pallet::error]
+    pub enum Error<T> {
+        LotteryNotStarted,
```
Same here, comments are missing. Once error happens, we can see some more details from blocks(polkadot-js)

Garandor replied:
Thx Jamie, added the comments and removed a few unused error types

Dengjianping commented on pallets/pallet-lottery/src/lib.rs:
```diff
@@ -0,0 +1,1077 @@
+                            .ok_or(Error::<T>::ArithmeticUnderflow)?
+                        {
+                            new_balance if new_balance.is_zero() => {
+                                TotalUsers::<T>::mutate(|users| *users -= 1);
```
`-` seems kind of unsafe to me here.

Apokalip replied:
Shouldnt be an issue, if this is called, users would be obviously 1 or more

Garandor replied:
TotalUsers is only used as a weight hint anyway, no code logic is impacted by this.

Can't hurt to do it properly tho, changed

Dengjianping commented on pallets/pallet-lottery/src/lib.rs:
```diff
@@ -0,0 +1,1077 @@
+            .or_else(|_| {
+                // Withdrawal needs extra collators to unstake to have enough funds to serve withdrawals, do it
+                let reserve = RemainingUnstakingBalance::<T>::get();
+                let mut remaining_to_withdraw = amount - reserve;
```
is it possible that `remaining_to_withdraw` is negative? I found `amount` is user's input.
Maybe I ask a stupid question.

Garandor replied:
yes it is possible but this case is handled in the first branch starting L464 so it can't happen in this `or_else` branch

Dengjianping commented on pallets/pallet-lottery/src/lib.rs:
```diff
@@ -0,0 +1,1077 @@
+            })?;
+            // END UNSTAKING SECTION
+            Self::deposit_event(Event::ScheduledWithdraw(caller, amount));
+            Ok::<(), DispatchError>(())
```
`Ok(().into())` should work here?

Garandor replied:
thx, changed, i had type resolution issues in earlier versions of the code

Dengjianping commented on pallets/pallet-lottery/src/lib.rs:
```diff
@@ -0,0 +1,1077 @@
+                        &caller,
+                        winnings,
+                        KeepAlive,
+                    ) // NOTE: If the transfer fails, the TXN get rolled back and the winnings stay in the map for claiming later
```
Can you add a event here?

Garandor replied:
oh yeah thanks, sounds like a good idea.
also this reminded me i created the Event type for `Withdrawn` but forgot to actually deposit it :sweat_smile:

Dengjianping commented on pallets/pallet-lottery/src/lib.rs:
```diff
@@ -0,0 +1,1077 @@
+            );
+            let lottery_drawing_call: CallOf<T> = Call::draw_lottery {}.into();
+            T::Scheduler::schedule_named(
+                Self::lottery_schedule_id(),
```
I'm not sure we should use the same name for every scheduled task, because if we use the same name, the previous scheduled task storage will be overwritten.

Garandor replied:
there can not be multiple lotteries running in parallel. This usecase is simply not supported for now

ghzlatarev replied:
Can we add this as a comment

Dengjianping commented on pallets/pallet-lottery/src/lib.rs:
```diff
@@ -0,0 +1,1077 @@
+        fn select_winning_balance(
+            max_winning_balance: BalanceOf<T>,
+        ) -> Result<BalanceOf<T>, Error<T>> {
+            const MAX_NUMBER_OF_RESAMPLES: u8 = 3;
```
constant should not be here.

Garandor replied:
it's a constant that only impacts this `fn` and isnt needed anywhere else. where would you put it. i think its fine there

Dengjianping commented on pallets/pallet-lottery/src/lib.rs:
```diff
@@ -0,0 +1,1077 @@
+            }
+        }
+
+        #[named]
```
Interesting, I never see such macro before.

Garandor replied:
it enables `log::trace!(function_name!());`

Commit:
fix randomness copyright headers

Signed-off-by: Adam Reif <Garandor@manta.network>

ghzlatarev commented on runtime/manta/src/lib.rs:
```diff
@@ -331,6 +336,81 @@ impl pallet_timestamp::Config for Runtime {
+        }
+        relay_chain_state_proof()
+            .read_optional_entry(
+                cumulus_primitives_core::relay_chain::well_known_keys::TWO_EPOCHS_AGO_RANDOMNESS,
```
can you put a comment why two epochs ago

Garandor replied:
done - and changed to 1 epoch ago

ghzlatarev commented on primitives/session-keys/src/inherent.rs:
```diff
@@ -0,0 +1,80 @@
+#[cfg_attr(feature = "std", derive(Debug, Decode))]
+/// TODO document
+pub enum InherentError {
+    /// TODO document
```
todo: docs

Garandor replied:
done

ghzlatarev commented on pallets/pallet-lottery/src/lib.rs:
```diff
@@ -0,0 +1,1077 @@
+    impl<T: Config> Default for GenesisConfig<T> {
+        fn default() -> Self {
+            Self {
+                min_deposit: 1u32.into(),
+                min_withdraw: 1u32.into(),
+                gas_reserve: 10_000u32.into(),
```
we should log a task to not forget to update these once we launch on both Manta and Calamari

Garandor replied:
thanks Georgi, this reminded me that we need to have extrinsics that can actually set these in the pallet lol

ghzlatarev commented on pallets/pallet-lottery/src/lib.rs:
```diff
@@ -0,0 +1,1077 @@
+            let winning_balance = Self::select_winning_balance(Self::total_pot())?;
+            let mut winner: Option<T::AccountId> = None;
+            let mut count: BalanceOf<T> = 0u32.into();
+            for (account, balance) in ActiveBalancePerUser::<T>::iter() {
```
we should make this storage item bounded because of this loop ActiveBalancePerUser

Garandor replied:
that would bound the number of users, doesn't seem desirable.
If this needs to be limited then it should be done by the "min_deposit" configurable

ghzlatarev replied:
yeah you're right the min-deposit should help with this. We should maybe add a ci to check that the min-deposit doesn't get something absurdly cheap though

Garandor replied:
@Apokalip could you help with implementing this pls?

Actually, this doesnt need to be done now. It's ok to log a followup

ghzlatarev commented on pallets/pallet-lottery/src/lib.rs:
```diff
@@ -0,0 +1,1077 @@
+
+        /// Stops the ongoing lottery and cancels the scheduled and any future drawings.
+        ///
+        /// This function cancels the scheduled drawing and cleans up bookkeeping.
```
How does it cleanup bookkeeping ? Looks like it just cancels the draw

Garandor replied:
good point, the comment was old and misleading. a stopped lottery is now just that, a lottery that does not have drawings but all other operations continue as normal (deposit/withdraw/rewards balance accrual)

Garandor replied:
updated comment, resolving

ghzlatarev commented on pallets/pallet-lottery/src/lib.rs:
```diff
@@ -0,0 +1,1077 @@
+                        left_overs.push((*request).clone());
+                        continue;
+                    }
+                    // stop paying people if we've run out of free funds.
+                    // The assumption is the collators serving these requests will
+                    // finish unstaking next round ( next lottery drawing )
```
this doesn't stop paying though, rather continues.

ghzlatarev replied:
is this edge case unit tested ?

Garandor replied:
`continue` is semantically identical with "stop paying".
there may be later requests of lower balance later in the list

ghzlatarev replied:
oh you mean stop paying the current, ok

ghzlatarev commented on pallets/pallet-lottery/src/lib.rs:
```diff
@@ -0,0 +1,1077 @@
+                    // stop paying people if we've run out of free funds.
+                    // The assumption is the collators serving these requests will
+                    // finish unstaking next round ( next lottery drawing )
+                    if request.balance > funds_available_to_withdraw {
```
we should update funds_available_to_withdraw after each successful transfer below

Garandor replied:
thanks for this good catch. this was a fairly large oversight. moved available_balance into the loop

ghzlatarev commented on pallets/pallet-lottery/src/lib.rs:
```diff
@@ -0,0 +1,1077 @@
+            <WithdrawalRequestQueue<T>>::mutate(|request_vec| -> Result<(), DispatchError> {
+                let mut left_overs: Vec<Request<_, _, _>> = Vec::new();
+                for request in request_vec.iter() {
+                    // Don't pay anyone still timelocked
+                    if now < request.block + <T as Config>::UnstakeLockTime::get() {
+                        left_overs.push((*request).clone());
+                        continue;
+                    }
+                    // stop paying people if we've run out of free funds.
+                    // The assumption is the collators serving these requests will
+                    // finish unstaking next round ( next lottery drawing )
+                    if request.balance > funds_available_to_withdraw {
+                        left_overs.push((*request).clone());
+                        continue;
+                    }
+                    // we know we can pay this out, do it
+                    <SumOfDeposits<T>>::mutate(|sum| (*sum).saturating_sub(request.balance));
+                    log::debug!(
+                        "Transferring {:?} to {:?}",
+                        request.balance.clone(),
+                        request.user.clone()
+                    );
+                    <T as pallet_parachain_staking::Config>::Currency::transfer(
+                        &Self::account_id(),
+                        &request.user,
+                        request.balance,
+                        KeepAlive,
+                    )?;
+                }
+                log::debug!(
+                    "Have {:?} requests left over after transfers",
+                    left_overs.len()
+                );
+                // Update T::WithdrawalRequestQueue if we paid at least one guy
+                if left_overs.len() != (*request_vec).len() {
+                    request_vec.clear();
+                    for c in left_overs {
+                        request_vec.push(c);
+                    }
+                }
```
if left_overs.len() != (*request_vec).len() {
    request_vec.clear();
    request_vec.append(&mut left_overs);
}

Garandor replied:
that seems clearer, thx!

ghzlatarev commented on pallets/pallet-lottery/src/tests.rs:
We can add more error case unit tests. Some simple ones like trying to withdraw more than deposited and trying to withdraw without having any deposit

ghzlatarev replied:
Or depositing less than the min

Garandor replied:
this was not done due to time constraints. I agree

ghzlatarev replied:
Yeah totally, let's just log it. Someone else could impl to get more familiar with the code.

Garandor replied:
https://github.com/Manta-Network/Manta/issues/1184

Commit:
fix: available funds not updated after withdrawal

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix misleading comment

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix: Only request_withdraw should decrease TotalPot, not collator staking operations

also, mutate is dangerous because rust doesn't actually check if the value was really written in the closure

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix: correct tracking of remaining_unstaking_balance

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix: delegation count + 1 on deposit

Signed-off-by: Adam Reif <Garandor@manta.network>

ghzlatarev commented on pallets/randomness/Cargo.toml:
adding vrf could be worth prioritizing in the near future

Garandor replied:
i see no action item here, marking resolved.

ghzlatarev replied:
We could log it, that's what i meant. Seems it like it could be worth spending time on it in the future ?

Commit:
add `set_min_deposit`, `set_min_withdraw` and `set_gas_reserve` extrinsics

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
pallet metadata

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
cleanup

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
add `Withdrawn` event

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
add Event::Claimed(account,balance)

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
document and remove unused errors

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix `surplus_funds` bookkeeping, fix restaking logic

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Safe math instead of -=

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
minor syntax

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
more verbose warning log

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Log failure for unavailable randomness instead of panicking

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Use ONE_EPOCH_AGO_RANDOMNESS and comment

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
clippy

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Merge branch 'manta' into jumboshrimps

Signed-off-by: Adam Reif <Garandor@manta.network>

ghzlatarev commented on runtime/manta/src/lib.rs:
```diff
@@ -331,6 +336,85 @@ impl pallet_timestamp::Config for Runtime {
+        relay_chain_state_proof()
+            .read_optional_entry(
+                // We use randomness from one relaychain epoch ago (4hrs Polkadot, 1hr Kusama) in combination with a (longer) `DrawingFreezeout` in pallet lottery
+                // to prevent manuipulation of the winning set by participants of the lottery
```
nit: spelling (manuipulation)

ghzlatarev replied:
the link is kinda broken - doesn't lead to the risks section

Garandor replied:
thx, fixed

ghzlatarev commented on runtime/manta/src/lib.rs:
```diff
@@ -331,6 +336,85 @@ impl pallet_timestamp::Config for Runtime {
+            .read_optional_entry(
+                // We use randomness from one relaychain epoch ago (4hrs Polkadot, 1hr Kusama) in combination with a (longer) `DrawingFreezeout` in pallet lottery
+                // to prevent manuipulation of the winning set by participants of the lottery
+                // NOTE: We operate under the following assumption https://github.com/Manta-Network/Manta/blob/f7b1a74e0261cc6a73081a9a86f8cd87aef8ea27/pallets/randomness/README.md#L85,
```
`The only incentive to prevent the validator from skipping the block is the block rewards. So the randomness value is only economically safe if the value at stake is lower than a block reward.` Doesn't that mean that if the lottery is higher than a block reward, then the randomess is not good ?

Garandor replied:
it means the last validator of the epoch before the drawing gets two tickets instead of one because he can choose to skip producing his block if the resulting random seed is better for him.
- This is fairly low-impact
- I've been told not to worry about fairness of the randomness algorithm itself

ghzlatarev replied:
ahaa ok i see, was thinking something else

ghzlatarev commented on pallets/pallet-lottery/src/tests.rs:
```diff
@@ -0,0 +1,407 @@
+// 202
```
this line is broken

Garandor replied:
oh lol, fixed thx

ghzlatarev commented on pallets/pallet-lottery/src/lib.rs:
```diff
@@ -0,0 +1,1175 @@
+    /// NOTE: how much KMA to keep in the pallet for gas
+    /// This must be initialized at genesis, otherwise the pallet will run out of gas at the first drawing
+    #[pallet::storage]
+    #[pallet::getter(fn gas_reserve)]
+    pub(super) type GasReserve<T: Config> = StorageValue<_, BalanceOf<T>, ValueQuery>;
```
what actually reduces this ?

Garandor replied:
Nothing, it's a configurable constant

Commit:
fix comment link

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix copyright commend

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fmt

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Merge branch 'manta' into jumboshrimps

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix: `AlreadyDelegated` error, selecting already withdrawing collators for withdrawal. Large refactor of `staking.rs`

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
clippy

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix round-0 stake/withdraw behavior. Only consider previously rewarded collators for deposit

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
round starts at 1 not 0 and fix test

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix: restake all available unstaked funds
fix: set genesis config for mock to have a normal gas_reserve for testing
add two more testcases, comments

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
Merge branch 'manta' into jumboshrimps

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
bump for CI

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fmt

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
taplo fmt

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
minor uncommit file

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
uncmommit more mistakenly added files

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
more files

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
readd types.json

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
unformat types.json

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
new test clippy

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix typo

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
fix benchmark

Signed-off-by: Adam Reif <Garandor@manta.network>

Commit:
into

Signed-off-by: Adam Reif <Garandor@manta.network>

ghzlatarev commented on pallets/pallet-lottery/src/staking/mod.rs:
```diff
@@ -0,0 +1,344 @@
+            );
+            return vec![]; // NOTE: Ensure we do not continue with the partial withdrawal
+        }
+        // TODO: Add random fallback
```
is there an issue for this ?

Garandor replied:
i'll remove this, that comment was a spur-of-the-moment thing and is not actually necessary, the previous strategy cannot actually fail

ghzlatarev commented on pallets/pallet-lottery/src/lib.rs:
```diff
@@ -0,0 +1,1210 @@
+                Error::<T>::NotImplemented.into(),
+            ))
+
+            // TODO: Unstake all collators, schedule return of all user deposits
+            // for collator in collators_we_staked_to {
+            //     do_unstake(collator);
+            // }
+            // TODO: Lock everything until this process is finished
+
+            // TODO: return user deposits and pay out winnings, deposit event
+
+            // Ok(())
```
issue for this

Garandor replied:
#1186

ghzlatarev commented on pallets/pallet-lottery/src/lib.rs:
```diff
@@ -0,0 +1,1210 @@
+                {
+                    random = T::RandomnessSource::random(&[n; 1]);
+                    log::debug!("select-winner using randomness {:?}", random);
+                    // TODO: The following check needs a change to pallet randomness but is static,
+                    //       so this can be done manually on deployment of the pallet
+                    // ensure!(
+                    //     random.1 = randomness_established_at_block
+                    //         .saturating_add(<T as Config>::DrawingFreezeout::get())
+                    //         < <frame_system::Pallet<T>>::block_number(),
+                    //     Error::<T>::PalletMisconfigured
+                    // );
```
issue for this ?

Garandor replied:
#1185

Apokalip reviewed (dismissed):
LGMT! Just add issues for all TODOs so we don't forget them

Commit:
cleanup comment

Signed-off-by: Adam Reif <Garandor@manta.network>
